### ISSUE TITLE
[BUG] cluster stuck in `Waiting for viable init node` after accidentally scaling down to 0

### ISSUE THREAD

**Issue by mateuszdrab** [rancher/rancher#39457]
**Rancher Server Setup**
- Rancher version: 2.6.9
- Installation option (Docker install/Helm Chart): Helm
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): k3s rancher-2.6.9 
- Proxy/Cert Details: N/A

**Information about the Cluster**
- Kubernetes version: 1.24.4
- Cluster Type (Local/Downstream): Downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): vSphere Infrastructure Provider RKE2

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions:

**Describe the bug**
I experienced a condition when provisioned nodes didn't request DHCP lease renewal and after a reboot, got new IPs.
The cluster fell apart and I attempted to fix this by scaling down the nodes to 1; however, as Rancher took a long time to act on my request I hit scale down again and end up with 0 desired nodes. I stopped Rancher quickly before it deleted the last node and grabbed an ETCD backup. I cleaned up the resources and started it back, scaled nodes up to 1. 
The newly created node has been stuck in Pending state and attempts to remove it and recreate would result in the same.

Rancher logs indicate CAPI being unable to connect to the CP of the cluster, which is expected in this case:

```
2022/10/31 16:03:22 [INFO] Handling backend connection request [stv-cluster-c-m-kxhpdc2x]
2022/10/31 16:03:27 [INFO] Handling backend connection request [c-m-kxhpdc2x]
I1031 16:03:28.979513      33 trace.go:205] Trace[851196838]: "Reflector ListAndWatch" name:pkg/mod/github.com/rancher/client-go@v1.24.0-rancher1/tools/cache/reflector.go:168 (31-Oct-2022 16:03:18.960) (total time: 10018ms):
Trace[851196838]: ---"Objects listed" error:<nil> 10018ms (16:03:28.979)
Trace[851196838]: [10.018525531s] [10.018525531s] END
2022/10/31 16:03:29 [INFO] [snapshotbackpopulate] rkecluster fleet-default/k3st: processing configmap kube-system/k3s-etcd-snapshots
2022/10/31 16:03:36 [INFO] error in remotedialer server [400]: websocket: close 1006 (abnormal closure): unexpected EOF
2022/10/31 16:03:37 [ERROR] [CAPI] error watching nodes on target cluster: error creating client and cache for remote cluster: error creating dynamic rest mapper for remote cluster "fleet-default/storclus": Get "https://10.43.105.208/k8s/clusters/c-m-wmgs58sk/api?timeout=10s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2022/10/31 16:03:37 [ERROR] [CAPI] Reconciler error: error creating client and cache for remote cluster: error creating dynamic rest mapper for remote cluster "fleet-default/storclus": Get "https://10.43.105.208/k8s/clusters/c-m-wmgs58sk/api?timeout=10s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2022/10/31 16:03:41 [INFO] Handling backend connection request [stv-cluster-c-m-tpx5859f]
```

**Result**
cluster stuck in `Waiting for viable init node`

**Expected Result**
A 0 node condition should not be allowed to happen and be blocked by the webhook or it should only be allowed to happen if the restart of the cluster is achieved with a restoration of the latest etcd backup.

**Additional context**
I need help getting the node recreated so that I can restore etcd on it and rebuild the cluster.
For safety purpose, I took the rights from the vcenter account that the CPI/CSI uses as I've had a similar situation to this happen in the past and I lost the cluster then as well and once it rebuilt, it also wiped all my pvcs from vCenter. The account used by Rancher to access vCenter is not affected by this change.

Thanks

**Comment by mateuszdrab** [comment]
Resolved the issue by peeking into source code in file `cluster-api\internal\controllers\machine\machine_controller.go` and finding that I have to reset the `ControlPlaneInitializedCondition` in the cluster's `clusters.cluster.x-k8s.io` resource using the [kubectl krew edit-status](https://github.com/ulucinar/kubectl-edit-status) plugin.

I will leave the issue open for someone to evaluate whether it's a bug worth fixing.

**Comment by Oats87** [comment]
I can see this from both sides -- part of the issue with allowing a new init node to be elected when all were eliminated is we don't want to accidentally trash a cluster (or cause a split brain) if something happens to the CRs. 

For now, the workaround should be fine, i.e. resetting the controlplane initialized condition -- for what it's worth, the control plane WAS initialized, and I don't really expect that condition to ever go back to false.

**Comment by mateuszdrab** [comment]
> I can see this from both sides -- part of the issue with allowing a new init node to be elected when all were eliminated is we don't want to accidentally trash a cluster (or cause a split brain) if something happens to the CRs. 
> 
> 
> 
> For now, the workaround should be fine, i.e. resetting the controlplane initialized condition -- for what it's worth, the control plane WAS initialized, and I don't really expect that condition to ever go back to false.
> 
> 
> 
> 

I agree that it should never need to go to false, but Rancher allowed me to scale down to 0, destroying the cluster and that is something that needs to be prevented. 
It's the second time it happened but the first time I didn't know enough to dig into the code and resolve the issue so has to recreate the cluster.

**Comment by Oats87** [comment]
@mateuszdrab Did you scale to zero through the UI? We can probably put a check in the UI to prevent this, but I really don't think we want to make changes that can lead to "worst cases" by way of being too autonomous.

**Comment by mateuszdrab** [comment]
> @mateuszdrab Did you scale to zero through the UI? We can probably put a check in the UI to prevent this, but I really don't think we want to make changes that can lead to "worse cases" by way of being too autonomous. 

Yes I believe the UI where the machine pool is blocks from clicking - when there's only one node but it's still possible to click on the node context menu and click scale down.
Or maybe it was due to the delay it took for rancher to show a node is being deleted after clicking scale down. Sometimes it seconds if not a minute, then you click again and suddenly things update and you realize it's too late.
I'm guessing this check needs to be implemented in the webhook or inside rancher itself to perhaps reject scaling down when there would be no nodes left even if the manifest says 0.

**Comment by Oats87** [comment]
There is a valid case in which a user may want to scale back down to zero temporarily, namely in the case of disaster recovery where they are looking to restore a cluster using an etcd snapshot. 

This somewhat ties into https://github.com/rancher/rancher/issues/41080 as an issue, where I am enhancing the provisioning framework to specifically force an etcd snapshot restore if the cluster goes down and there are no functional etcd nodes left.

We can keep this issue open, but I am against enforcing allowing the scale count to be set to 0.

@rancher/rancher-team-2-hostbusters-dev does anyone on the team have any thoughts about this?
